### DISCUSSION TITLE
jupyter lab基本操作

### DISCUSSION THREAD

**Discussion by enoch-lightning** [1101-GS4719-C/Python-Community-GS4719-C#242] [General]
本學期使用jupyter lab與Visual Studio Code這兩種程式(還是「軟體」、「界面」、「命令提示字元」、「終端機」或「整合式開發環境=整合開發環境(integrated development environment)」？)去「撰寫=寫=打」Python這種程式。
在此，先整理出jupyter lab的操作步驟。
- 第一步：下載jupyter lab
- 第二步：安裝jupyter lab
- 第三步：打開「命令提示字元(command line interface, cmd)」，在某一個「目錄=資料夾」中(還是下？這有沒有什麼權限或層級的關係？)，使用電腦的半形英文輸入法輸入jupyter lab或jupyter-lab，再按下enter來執行jupyter-lab.exe
- 第四步：新增jupyter notebook開始撰寫Python程式
- 第五步：撰寫完成後，先關掉jupyter lab「視窗=頁籤」
- 第六步：回到命令提示字元，按下ctrl+c。命令提示字元會顯示警告(warning)訊息。這些警告訊息我一個都看不懂。
[W 2021-12-29 22:19:33.434 LabApp] Could not determine jupyterlab build status without nodejs
[W 2021-12-29 22:19:34.137 ServerApp] Notebook Untitled2.ipynb is not trusted
[I 2021-12-29 22:19:34.295 ServerApp] Kernel started: 788b4bb2-2173-4f1a-93dd-6ca4c6d98122
[I 2021-12-29 22:19:34.819 ServerApp] Kernel started: 00280568-60d3-45c7-a58e-d88429f52ff5
[I 2021-12-29 22:19:58.294 ServerApp] Interrupted...
[IPKernelApp] WARNING | Parent appears to have exited, shutting down.
[IPKernelApp] WARNING | Parent appears to have exited, shutting down.
同時，jupyter lab會顯示如下的畫面
![image](https://user-images.githubusercontent.com/86505454/147671918-92eb9017-f446-47f9-b5d7-bcd390f95a86.png)
這時，只有兩個選擇
1.用滑鼠左鍵/觸控板左邊按Dismiss一下
2.將滑鼠移到右上角的x，用滑鼠左鍵/觸控板左邊按一下，關掉此jupyter lab視窗
- 第七步：回到命令提示字元，再按ctrl+c一次，就會回到只顯示資料夾工作目錄的情況。
`C:\Users\enoch>`
- 第八步：在命令提示字元輸入exit或exit()，再按下enter，便可離開命令提示字元

**Comment by enoch-lightning** [comment]
# 在jupyer lab裡寫程式
- 換下一行=換行=按電腦鍵盤的enter一下
- ctrl+enter=執行這個cell
- shift+enter=執行這個cell，在執行完這個cell後，在下面新增一個cell

**Comment by enoch-lightning** [comment]
# 可以直接找到jupyter lab的執行檔，並直接以滑鼠左鍵點擊兩下開啟執行
[按jupyter lab執行檔執行jupyter lab](https://youtu.be/EAP16YnXaSU)

**Comment by 107602505** [comment]
如果是直接google 進入https://jupyter.org/ 這個網頁可以用瀏覽器使用jupyter lab，如果是從命令提示字元打jupyter lab進入的話也可以，而且每次打開它都會保存之前打過的指令，有利於複習一些上課老師教的指令，只不過要注意在撰寫程式的時候不要將cmd關掉就好。

**Reply by enoch-lightning** [reply]
你說的這個網頁，就是老師之前在 [#183](https://github.com/NCU-GS4719-Python/Python-Community/discussions/183)提到的網頁。只是可以用自己的電腦操作，也可以用手機或平板操作。因為真正在跑程式的地方，是jupyter官網提供的電腦。我想這種電腦可能就是「伺服器=server」，我們藉由Binder這個API與jupyter的伺服器互動。Binder將我想使用jupyter伺服器中的jupyter lab的需求傳給jupyter的伺服器，jupyter伺服器透過Binder將jupyter lab的連線「回傳=傳回=return」給我，讓我可以使用jupyter lab。
- 伺服器(server)本身是沒有螢幕(screen)的，因此伺服器的標準輸出是磁碟，包含C:(C磁碟=C磁碟機)與D:(D磁碟=D磁碟機)與硬碟(我不太清楚磁碟與硬碟是不是同一個東西)
- API：**a**pplication **p**rogramming **i**nterface，取三個英文單字的a、p與i就形成API。application programming interface的中文是「應用程式界面」

**Comment by enoch-lightning** [comment]
# 在jupyter lab進入無限迴圈怎麼辦
在jupyter lab如果盡入無限迴圈，必須重新啟動(restart)Python直譯器(kernel)，才能再度正常編寫程式。
## 重新啟動Python直譯器

1. 切斷jupyter lab與Python直譯器的連接
2. 清除RAM這種揮發性記憶體中，有關Python的資料。如整數物件1。
3. 再度將jupyter lab與Python直譯器連接
4. 重新正常編寫程式
[jupyter lab restart kernel](https://youtu.be/hUaD-dV6tBs)

**Reply by 107602505** [reply]
之前沒遇到這種情況但感覺難免會碰到，感謝原PO提出這個問題和解決辦法!

**Comment by ghost** [comment]
jupyter lab跟jupyter notebook不知道有沒有人知道差在哪?

**Reply by enoch-lightning** [reply]
我目前的理解如下：jupyter lab與jupyter notebook的功能完全一樣，連「熱鍵(快捷鍵)」都一樣。差別就是在於操作介面的長相不一樣。像是jupyter lab在左方可以看到目前的工作目錄下(裡面)有什麼，方便我在打Python程式的時候，看到我可以在工作目錄下使用什麼檔案，我新增，或刪除什麼檔案。但使用jupyter notebook只看得到我寫Python這種程式的部份。

**Reply by chia-kai-chang** [reply]
JupyterLab可以想像成升級版的Jupyter Notebook，提供多標簽頁支援，線上終端和文件管理等一系列方便的功能，接近於一個線上版的Python IDE。

**Reply by 107602505** [reply]
Colab 筆記本也是一個可在瀏覽器上撰寫及執行Python程式碼和 RTF 格式，並附帶圖片、HTML、LaTeX 等其他格式的內容，且具備下列優點：不必進行任何設定、免費使用 GPU、還可以和朋友輕鬆共用

**Comment by enoch-lightning** [comment]
# jupyter lab的自動偵測縮排功能
## 以while迴圈為例
下面這部影片最初的目的在介紹jupyter lab的自動偵測縮排功能，經由一些實驗，得出以下結論。

1. Python的同一程式區塊，只要縮排的長度一樣長，就可以執行。
2. 在打Python程式的時候，1tab=4空白；若想把'\t'就是tab與「空白」使用print()函式列印輸出到標準輸出，也就是螢幕上，1tab=8空白。
3. 在jupyter lab與jupyter notebook中，縮排的產生可以將tab與空白混合使用。
4. 老師建議使用四個空格縮排，應該只是統一習慣，沒有特別目的。
[影片：jupyter lab自動偵測縮排](https://youtu.be/sUkVxzzyuhc)
